fflas-ffpack/fflas/fflas_faxpy.inl

```diff
@@ -27,18 +28,168 @@
+    axpyp(const Field &F, const typename Field::Element a, typename Field::ConstElement_ptr X, typename Field::Element_ptr Y, const size_t n, HelperMod<Field> &H)
+    {
+        size_t i = 0;
+
+        for (; i < n ; i++)
+        {
+            Y[i] = reduce(a*X[i] + Y[i], H);
+        }
+    }
+
+
+    template<class Field>
+    inline typename std::enable_if<FFLAS::support_fast_mod<typename Field::Element>::value, void>::type
+    axpyp(const Field &F, const typename Field::Element a, typename Field::ConstElement_ptr X, typename Field::Element_ptr Y, const size_t n, const size_t incX, const size_t incY,
+              HelperMod<Field> &H)
+    {
+        typename Field::ConstElement_ptr Xi = X;
+        typename Field::Element_ptr Yi=Y;
+        for (; Xi < X+n*incX; Xi+=incX, Yi+=incY )
+            *Yi = reduce(a*(*Xi) + (*Yi), H);
+    }
+
+} // unswitch
+} // vectorised
+} // FFLAS
+
+namespace FFLAS { namespace vectorised {
+
+    // simd_mod => fast_mod, so this declaration's one in two
+    template<class Field>
+    inline typename std::enable_if<FFLAS::support_fast_mod<typename Field::Element>::value, void>::type
+    axpyp(const Field &F, const typename Field::Element a, typename Field::ConstElement_ptr X, typename Field::Element_ptr Y, const size_t n)
+    {
+        HelperMod<Field> H(F);
+
+        unswitch::axpyp(F,a,X,Y,n,H);
+    }
+
+    template<class Field>
+    inline typename std::enable_if<FFLAS::support_fast_mod<typename Field::Element>::value, void>::type
+    axpyp(const Field &F, const typename Field::Element a, typename Field::ConstElement_ptr X, typename Field::Element_ptr Y, const size_t n, const size_t incX, const size_t incY)
+    {
+        HelperMod<Field> H(F);
+
+        unswitch::axpyp(F,a,X,Y,n,incX,incY,H);
+    }
+
+} // vectorised
+} // FFLAS
+
+namespace FFLAS { namespace details {
+
+    /*** Entry-points for specialised code ***/
+    // simd_mod => fast_mod;
+    // will default to the best supported option in unswitch
+
+    template<class Field>
+    inline typename std::enable_if<FFLAS::support_fast_mod<typename Field::Element>::value, void>::type
+    faxpy( const Field & F , const size_t N,
            const typename Field::Element a,
            typename Field::ConstElement_ptr X, const size_t incX,
-           typename Field::Element_ptr Y, const size_t incY )
+           typename Field::Element_ptr Y, const size_t incY,
+           FieldCategories::ModularTag)
     {
+        if((incX == 1) && (incY == 1))
+        {
+            vectorised::axpyp(F,a,X,Y,N);
+        }
+        else
+        {
+            if (N < FFLASFFPACK_COPY_REDUCE) // defined in fflas_freduce.inl
+            {
+                vectorised::axpyp(F,a,X,Y,N,incX,incY);
+            }
+            else
+            {
+                typename Field::Element_ptr Xc = fflas_new (F,N) ;
```

Comment:
If only one of X or Y has an inc !=1 then this code copies both vectors anyway!

Reply:
This looks indeed wasteful

Reply:
Should be fixed in latest commit